src/viewport-observer.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {devAssert} from '../src/log';
+import {isIframed} from './dom';
+
+/**
+ * Returns an IntersectionObserver tracking the Viewport.
+ * Only use this if x-origin iframe rootMargin support is considered nice-to-have,
+ * since if not supported this InOb will fallback to one without rootMargin.
+ *
+ * - If iframed: rootMargin is ignored unless natively supported (Chrome 81+).
+ * - If not iframed: all features work properly in both polyfill and built-in.
+ *
+ * @param {function(!Array<!IntersectionObserverEntry>)} ioCallback
+ * @param {!Window} win
+ * @param {number=|!Array<number>=} threshold
+ *
+ * @return {!IntersectionObserver}
+ */
+export function createViewportObserver(ioCallback, win, threshold) {
+  const iframed = isIframed(win);
+  const root = /** @type {?Element} */ (iframed
+    ? /** @type {*} */ (win.document)
+    : null);
+
+  // TODO(#30794): See if we can safely remove rootMargin without adversely
+  // affecting metrics.
+  const rootMargin = '25%';
+
+  try {
+    return new win.IntersectionObserver(ioCallback, {
+      root,
+      rootMargin,
+      threshold,
+    });
+  } catch (e) {
+    return new win.IntersectionObserver(ioCallback, {rootMargin, threshold});
+  }
+}
+
+/** @type {!WeakMap<!Window, !IntersectionObserver>} */
+const viewportObservers = new WeakMap();
+
+/** @type {!WeakMap<!Element, function(boolean)>} */
+const viewportCallbacks = new WeakMap();
+
+/**
+ * Lazily creates an IntersectionObserver per Window to track when elements
+ * enter and exit the viewport. Fires viewportCallback when this happens.
+ *
+ * @param {!Element} element
+ * @param {function(boolean)} viewportCallback
+ */
+export function observe(element, viewportCallback) {
+  // There should never be two unique observers of the same element.
+  devAssert(
+    !viewportCallbacks.has(element) ||
+      viewportCallbacks.get(element) === viewportCallback
```

Comment:
Nit: These won't be eliminated in production.

Reply:
I thought devAssert was always stripped?

Reply:
sorry if I'm being dense here but can you explain more?

1. why won't this be DCE'd
2. how can i fix it

Reply:
answer: put it all behind an if condition for `getMode().localDev`